File: 0901-1000/921-930/930M_Binary_Subarrays_With_Sum/java/main.java
// 930. Binary Subarrays With Sum
// https://leetcode.com/problems/binary-subarrays-with-sum/
// T.C.: O(n)
// S.C.: O(1)

// java main.java
class Main {

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[] nums = { 1, 0, 1, 0, 1 };
    int expected = 4;
    int result = solution.numSubarraysWithSum(nums, 2);
    System.out.println(result + " " + (expected == result ? "true" : "false"));

    nums = new int[] { 0, 0, 0, 0, 0 };
    expected = 15;
    result = solution.numSubarraysWithSum(nums, 0);
    System.out.println(result + " " + (expected == result ? "true" : "false"));

    nums = new int[] { 1, 0, 0, 1, 0, 1 };
    expected = 5;
    result = solution.numSubarraysWithSum(nums, 2);
    System.out.println(result + " " + (expected == result ? "true" : "false"));
  }
}

class Solution {

  public int numSubarraysWithSum(int[] nums, int goal) {
    int currentCount = 0;
    int left = 0;
    int leftSum = 0;
    int right = 0;
    int rightSum = 0;
    for (int end = 0; end < nums.length; end++) {
      leftSum += nums[end];
      while (left < end && leftSum > goal) {
        leftSum -= nums[left++];
      }

      rightSum += nums[end];
      while (right < end && (rightSum > goal || (rightSum == goal && nums[right] == 0))) {
        rightSum -= nums[right++];
      }

      if (leftSum == goal) {
        currentCount += right - left + 1;
      }
    }

    return currentCount;
  }
}

File: 0901-1000/921-930/930M_Binary_Subarrays_With_Sum/rust/src/main.rs
// 930. Binary Subarrays With Sum
// https://leetcode.com/problems/binary-subarrays-with-sum/
// T.C.: O(n)
// S.C.: O(1)
use colored::Colorize;
use colored::ColoredString;

struct Solution;

impl Solution {
    pub fn num_subarrays_with_sum(nums: Vec<i32>, goal: i32) -> i32 {
        let mut current_count = 0;
        let mut left_sum = 0;
        let mut right_sum = 0;
        let mut left = 0;
        let mut right = 0;
        for end in 0..nums.len() {
            left_sum += nums[end];
            while left < end && left_sum > goal {
                left_sum -= nums[left];
                left += 1;
            }

            right_sum += nums[end];
            while right < end && (right_sum > goal || (right_sum == goal && nums[right] == 0)) {
                right_sum -= nums[right];
                right += 1;
            }

            if left_sum == goal {
                current_count += right - left + 1;
            }
        }
        return current_count as i32;
    }
}

fn main() {
    let mut nums = vec![1, 0, 1, 0, 1];
    test(nums, 2, 4);

    nums = vec![0, 0, 0, 0, 0];
    test(nums, 0, 15);

    nums = vec![1, 0, 0, 1, 0, 1];
    test(nums, 2, 5);
}

fn test(nums: Vec<i32>, goal: i32, expected: i32) {
    let result = Solution::num_subarrays_with_sum(nums.clone(), goal);

    println!("\nresult: {}\nassert: {}", number_colored(result), bool_colored(result == expected));
}

fn bool_colored(value: bool) -> ColoredString {
    let text;
    if value {
        text = value.to_string().green();
    } else {
        text = value.to_string().red();
    }
    return text;
}

fn number_colored(value: i32) -> ColoredString {
    return value.to_string().yellow();
}
